### PR TITLE
Store numeric tags as metrics

### PR DESCRIPTION
**Pull request by delner**
In order to support numeric facets in the interim, this pull request changes span behavior to save all numeric tags as metrics on the span. Keys between tags and metrics are also unique, meaning you can't have a tag and metric of the same name on the same span.

### CODE REVIEW

**Comment by delner** [comment]
Ran some tests to verify numeric tags show up in the UI as we expect; that seems to be behaving correctly. Status codes, ports, process IDs, etc are still visible.

Unit tests are all that remain.

**Comment by delner** [comment]
Okay it's become clear from the unit tests although setting tags as metrics is fine, it has side effects for `get_tag`: many numeric tags no longer any value for `get_tag`. We can change `get_tag` to return a metric, and `get_metric` to return a tag, but it's not clear whether this is correct.

**Comment by marcotc** [comment]
Given that tags and metrics will share the same domain for their keys, having tags return metrics or metrics return tags for the same key does seem reasonable at first glance.

**Comment by delner** [comment]
Updated this to have `get_tag` and `get_metric` to return both types; had to update a lot of tests though to make this work.

Should be ready for final review.

File: lib/ddtrace/span.rb [hunk 1]
```diff
@@ -92,12 +102,15 @@ def clear_tag(key)
 
     # Return the tag with the given key, nil if it doesn't exist.
     def get_tag(key)
-      @meta[key]
+      @meta[key] || @metrics[key]
```
Comment (brettlangdon) [inline lib/ddtrace/span.rb]: this is so weird to read for me, in Python empty string is falsey

File: spec/ddtrace/contrib/rails/redis_cache_spec.rb [hunk 2]
```diff
@@ -40,7 +40,12 @@
   end
 
   let(:cache_store_name) do
-    Gem.loaded_specs['redis-activesupport'] ? 'redis_store' : 'redis_cache_store'
+    if Gem.loaded_specs['redis-activesupport'] \
```
Comment (brettlangdon) [inline spec/ddtrace/contrib/rails/redis_cache_spec.rb]: this change seems unrelated?

Comment (delner) [reply spec/ddtrace/contrib/rails/redis_cache_spec.rb]: No, tests were breaking. No idea how these were passing before but we changed the default store for Rails 5+ to use the internal Redis implementation instead of the gem.

File: lib/ddtrace/span.rb (hunk 1 above)
Comment (delner) [reply lib/ddtrace/span.rb]: Only false values are `false` and `nil` in Ruby. Everything else evaluates to `true`.

**Comment by lraucy** [comment]
I don't see anything about the special case of `>= 2^53` numbers, am I missing it or did that requirement get lost?

Just to clarify this point, numbers greater than 2^53 (in absolute value) must be put in `Meta`, otherwise we're facing a major regression with any customer putting IDs in those tags, as those numbers will be rounded because of the underlying float representation of `span.Metrics`.

**Comment by delner** [comment]
@lraucy Added in the size limit.

File: spec/ddtrace/contrib/rails/redis_cache_spec.rb (hunk 2 above)
Comment (brettlangdon) [reply spec/ddtrace/contrib/rails/redis_cache_spec.rb]: hmm, weird if it is only failing on this branch, otherwise if unrelated to your changes it should probably be in a separate PR

File: lib/ddtrace/span.rb [hunk 3]
```diff
@@ -31,6 +31,9 @@ class Span
     # parsing 64-bit integers for distributed tracing since an upstream system may generate one
     EXTERNAL_MAX_ID = 2**64
 
+    # This limit is for numeric tags because uint64 could end up rounded.
+    MAX_NUMERIC_TAG_SIZE = 2**53
```
Comment (marcotc) [inline lib/ddtrace/span.rb]: Does this limit also apply to floating-point numbers greater than 2⁵³? The comment makes me think that this is only a problem for integer numbers.

Comment (delner) [reply lib/ddtrace/span.rb]: Interestingly enough:

```
irb(main):005:0> ((2**53)+0.5) > 2**53
=> false
irb(main):006:0> ((2**53)+1.5) > 2**53
=> true
```

Comment (lraucy) [reply lib/ddtrace/span.rb]: Indeed this limit is scoped to integer as they're exact numbers. For floats (which are already stored as approximation), we can (and should?) still put them in `Metrics` field, as there is no loss of precision involved. Sorry for the lack of precision here.

Comment (delner) [reply lib/ddtrace/span.rb]: Okay, so all numbers are stored as metrics unless its an integer greater than 2^53.

File: lib/ddtrace/span.rb [hunk 4]
```diff
@@ -80,7 +83,19 @@ def initialize(tracer, name, options = {})
...
-      @meta[key] = value.to_s
+      # Keys must be unique between tags and metrics
+      @metrics.delete(key) if @metrics.key?(key)
+
+      # NOTE: Adding numeric tags as metrics is stop-gap support
+      #       for numeric typed tags. Eventually they will become
+      #       tags again.
+      # Any numeric that is not an integer greater than max size is logged as a metric.
+      # Everything else gets logged as a tag.
+      if value.is_a?(Numeric) && !(value.is_a?(Integer) && value > MAX_NUMERIC_TAG_SIZE)
```
Comment (marcotc) [inline lib/ddtrace/span.rb]: From @lraucy's comment:
> numbers greater than 2^53 (in absolute value) must be put in Meta

we also have to handle negative values smaller than MAX_NUMERIC_TAG_SIZE.

File: lib/ddtrace/span.rb [hunk 5]
```diff
@@ -31,6 +31,9 @@ class Span
     # parsing 64-bit integers for distributed tracing since an upstream system may generate one
     EXTERNAL_MAX_ID = 2**64
 
+    # This limit is for numeric tags because uint64 could end up rounded.
+    NUMERIC_TAG_SIZE_RANGE = ((-1 * 2**53)..2**53)
```
Comment (marcotc) [inline lib/ddtrace/span.rb]: Super minor, but I think this can be simplified like so:
```suggestion
    NUMERIC_TAG_SIZE_RANGE = (-2**53..2**53)
```

File: lib/ddtrace/span.rb [hunk 6]
```diff
@@ -80,7 +83,19 @@ def initialize(tracer, name, options = {})
     #
     #   span.set_tag('http.method', request.method)
     def set_tag(key, value = nil)
-      @meta[key] = value.to_s
+      # Keys must be unique between tags and metrics
+      @metrics.delete(key) if @metrics.key?(key)
```
Comment (marcotc) [inline lib/ddtrace/span.rb]: Do we need to check for key existence here?
I believe `Hash#delete` is safe on non-existing keys.

Same comment applies to `def set_metric(key, value)` method.

File: spec/ddtrace/contrib/rails/redis_cache_spec.rb [hunk 7]
```diff
@@ -40,7 +40,12 @@
...
+       && Gem::Version.new(Rails::VERSION::STRING) < Gem::Version.new('5.0')
```
Comment (marcotc) [inline spec/ddtrace/contrib/rails/redis_cache_spec.rb]: `redis_cache_store` was introduced in the minor release `5.2`: https://weblog.rubyonrails.org/2018/4/9/Rails-5-2-0-final/
```suggestion
       && Gem::Version.new(Rails::VERSION::STRING) < Gem::Version.new('5.2')
```

**Comment by delner** [comment]
Minor CI issue is breaking this build; will resolve elsewhere.
